Document (or fix) limitations of stateful testing

Because stateful testing is currently built on find, it had a number of features supported by `@given` that do not work. These include:

* Coverage guided testing (this is a particularly glaring omission because stateful testing is where this would be most useful!)
* Statistics
* (Most) health checks
* Deadlines

Currently this isn't documented anywhere. We should document this at a minimum, though better yet would be rebuilding stateful testing and given to have more of a common mechanism.

This is a minor limitation, but `@rule` and `@pytest.mark.parametrize` are not compatible (unlike `@given`). I didn't see this reported in a separate issue.

MWE:

```python
import pytest

from hypothesis import stateful
from hypothesis.strategies import booleans


class PytestMachine(stateful.RuleBasedStateMachine):

    @stateful.rule(i=booleans())
    @pytest.mark.parametrize('hi', (1, 2, 3))
    def parametrize_after_rule(self, hi, i):
        pass


TestPytestMachine = PytestMachine.TestCase
```

Failure: 

```
 test_compat.py::TestPytestMachine::runTest <- [...]/python3.6/site-packages/hypothesis/stateful.py  
[...]/python3.6/site-packages/hypothesis/stateful.py:192: in runTest
    run_state_machine_as_test(state_machine_class)
[...]/python3.6/site-packages/hypothesis/stateful.py:110: in run_state_machine_as_test
    breaker.run(state_machine_factory(), print_steps=True)
[...]/python3.6/site-packages/hypothesis/stateful.py:248: in run
    state_machine.execute_step(value)
[...]/python3.6/site-packages/hypothesis/stateful.py:589: in execute_step
    result = rule.function(self, **data)
E   TypeError: parametrize_after_rule() missing 1 required positional argument: 'hi'
---------------------------------- Hypothesis ----------------------------------
Step #1: parametrize_after_rule(i=False)

```

--- a/.isort.cfg
+++ b/.isort.cfg
@@ -1,5 +1,5 @@
 [settings]
-known_third_party = attr, click, dateutil, django, faker, flaky, numpy, pandas, pytz, pytest, scipy, unicodenazi
+known_third_party = attr, click, dateutil, django, faker, flaky, numpy, pandas, pytz, pytest, pyup, scipy, unicodenazi
 known_first_party = hypothesis
 add_imports = from __future__ import absolute_import, from __future__ import print_function, from __future__ import division
 multi_line_output = 3

--- a/.travis.yml
+++ b/.travis.yml
@@ -59,7 +59,8 @@ jobs:
       env: TASK=check-unicode
     - env: TASK=check-py27-typing
     - env: TASK=check-nose
-    - env: TASK=check-pytest30
+    # https://github.com/HypothesisWorks/hypothesis/issues/1702
+    # - env: TASK=check-pytest30
     - env: TASK=check-faker070
     - env: TASK=check-faker-latest
     - env: TASK=check-django21

--- a/hypothesis-python/RELEASE.rst
+++ b/hypothesis-python/RELEASE.rst
@@ -1,0 +1,15 @@
+RELEASE_TYPE: minor
+
+This release changes the stateful testing backend from
+:func:`~hypothesis.find` to use :func:`@given <hypothesis.given>`
+(:issue:`1300`).  This doesn't change how you create stateful tests,
+but does make them run more like other Hypothesis tests.
+
+:func:`@reproduce_failure <hypothesis.reproduce_failure>` and
+:func:`@seed <hypothesis.seed>` now work for stateful tests.
+
+Stateful tests now respect the :attr:`~hypothesis.settings.deadline`
+and :attr:`~hypothesis.settings.suppress_health_check` settings,
+though they are disabled by default.  You can enable them by using
+:func:`@settings(...) <hypothesis.settings>` as a class decorator
+with whatever arguments you prefer.

--- a/hypothesis-python/src/hypothesis/core.py
+++ b/hypothesis-python/src/hypothesis/core.py
@@ -71,7 +71,6 @@ from hypothesis.internal.compat import (
     hbytes,
     int_from_bytes,
     qualname,
-    str_to_bytes,
 )
 from hypothesis.internal.conjecture.data import ConjectureData, StopTest
 from hypothesis.internal.conjecture.engine import ConjectureRunner, ExitReason, sort_key
@@ -85,7 +84,6 @@ from hypothesis.internal.reflection import (
     arg_string,
     convert_positional_arguments,
     define_function_signature,
-    fully_qualified_name,
     function_digest,
     get_pretty_function_description,
     impersonate,
@@ -639,7 +637,7 @@ class StateForActualGivenExecution(object):
         # Tell pytest to omit the body of this function from tracebacks
         __tracebackhide__ = True
         if global_force_seed is None:
-            database_key = str_to_bytes(fully_qualified_name(self.test))
+            database_key = function_digest(self.test)
         else:
             database_key = None
         self.start_time = benchmark_time()

--- a/hypothesis-python/src/hypothesis/internal/reflection.py
+++ b/hypothesis-python/src/hypothesis/internal/reflection.py
@@ -93,6 +93,10 @@ def function_digest(function):
         hasher.update(str_to_bytes(repr(getfullargspec(function))))
     except TypeError:
         pass
+    try:
+        hasher.update(function._hypothesis_internal_add_digest)
+    except AttributeError:
+        pass
     return hasher.digest()
 
 

--- a/hypothesis-python/src/hypothesis/stateful.py
+++ b/hypothesis-python/src/hypothesis/stateful.py
@@ -27,33 +27,31 @@ execution to date.
 from __future__ import absolute_import, division, print_function
 
 import inspect
-import traceback
 from copy import copy
 from unittest import TestCase
 
 import attr
 
 import hypothesis.internal.conjecture.utils as cu
-from hypothesis._settings import Verbosity, note_deprecation, settings as Settings
-from hypothesis.control import BuildContext
-from hypothesis.core import EXCEPTIONS_TO_FAIL, find
-from hypothesis.errors import (
-    Flaky,
-    HypothesisException,
-    InvalidArgument,
-    InvalidDefinition,
-    NoSuchExample,
+import hypothesis.strategies as st
+from hypothesis._settings import (
+    HealthCheck,
+    Verbosity,
+    note_deprecation,
+    settings as Settings,
+    unlimited,
 )
+from hypothesis.control import current_build_context
+from hypothesis.core import given
+from hypothesis.errors import InvalidArgument, InvalidDefinition
 from hypothesis.internal.compat import int_to_bytes, string_types
-from hypothesis.internal.conjecture.data import StopTest
-from hypothesis.internal.conjecture.utils import calc_label_from_name, integer_range
-from hypothesis.internal.reflection import nicerepr, proxies
-from hypothesis.reporting import current_verbosity, report, verbose_report
+from hypothesis.internal.reflection import function_digest, nicerepr, proxies
+from hypothesis.internal.validation import check_type
+from hypothesis.reporting import current_verbosity, report
 from hypothesis.searchstrategy.strategies import OneOfStrategy, SearchStrategy
-from hypothesis.strategies import fixed_dictionaries, just, one_of, runner, tuples
 from hypothesis.vendor.pretty import CUnicodeIO, RepresentationPrinter
 
-STATE_MACHINE_RUN_LABEL = calc_label_from_name("another state machine step")
+STATE_MACHINE_RUN_LABEL = cu.calc_label_from_name("another state machine step")
 
 if False:
     from typing import Any, Dict, List, Text  # noqa
@@ -72,33 +70,6 @@ class TestCaseProperty(object):  # pragma: no cover
         raise AttributeError(u"Cannot delete TestCase")
 
 
-def find_breaking_runner(state_machine_factory, settings=None):
-    def is_breaking_run(runner):
-        try:
-            runner.run(state_machine_factory())
-            return False
-        except HypothesisException:
-            raise
-        except EXCEPTIONS_TO_FAIL:
-            verbose_report(traceback.format_exc)
-            return True
-
-    if settings is None:
-        try:
-            settings = state_machine_factory.TestCase.settings
-        except AttributeError:
-            settings = Settings.default
-
-    search_strategy = StateMachineSearchStrategy(settings)
-
-    return find(
-        search_strategy,
-        is_breaking_run,
-        settings=settings,
-        database_key=state_machine_factory.__name__.encode("utf-8"),
-    )
-
-
 def run_state_machine_as_test(state_machine_factory, settings=None):
     """Run a state machine definition as a test, either silently doing nothing
     or printing a minimal breaking program and raising an exception.
@@ -107,16 +78,50 @@ def run_state_machine_as_test(state_machine_factory, settings=None):
     GenericStateMachine when called with no arguments - it can be a class or a
     function. settings will be used to control the execution of the test.
     """
-    try:
-        breaker = find_breaking_runner(state_machine_factory, settings)
-    except NoSuchExample:
-        return
-    try:
-        with BuildContext(None, is_final=True):
-            breaker.run(state_machine_factory(), print_steps=True)
-    except StopTest:
-        pass
-    raise Flaky(u"Run failed initially but succeeded on a second try")
+    if settings is None:
+        try:
+            settings = state_machine_factory.TestCase.settings
+            check_type(Settings, settings, "state_machine_factory.TestCase.settings")
+        except AttributeError:
+            settings = Settings(
+                timeout=unlimited,
+                deadline=None,
+                suppress_health_check=HealthCheck.all(),
+            )
+    check_type(Settings, settings, "settings")
+
+    @settings
+    @given(st.data())
+    def run_state_machine(factory, data):
+        machine = factory()
+        check_type(GenericStateMachine, machine, "state_machine_factory()")
+        data.conjecture_data.hypothesis_runner = machine
+
+        n_steps = settings.stateful_step_count
+        should_continue = cu.many(
+            data.conjecture_data, min_size=1, max_size=n_steps, average_size=n_steps
+        )
+
+        print_steps = (
+            current_build_context().is_final or current_verbosity() >= Verbosity.debug
+        )
+        try:
+            if print_steps:
+                machine.print_start()
+            machine.check_invariants()
+
+            while should_continue.more():
+                value = data.conjecture_data.draw(machine.steps())
+                if print_steps:
+                    machine.print_step(value)
+                machine.execute_step(value)
+                machine.check_invariants()
+        finally:
+            if print_steps:
+                machine.print_end()
+            machine.teardown()
+
+    run_state_machine(state_machine_factory)
 
 
 class GenericStateMachineMeta(type):
@@ -158,8 +163,6 @@ class GenericStateMachine(
     And if this ever produces an error it will shrink it down to a small
     sequence of example choices demonstrating that.
     """
-
-    find_breaking_runner = None  # type: classmethod
 
     def steps(self):
         """Return a SearchStrategy instance the defines the available next
@@ -214,7 +217,11 @@ class GenericStateMachine(
             pass
 
         class StateMachineTestCase(TestCase):
-            settings = Settings()
+            settings = Settings(
+                timeout=unlimited,
+                deadline=None,
+                suppress_health_check=HealthCheck.all(),
+            )
 
         # We define this outside of the class and assign it because you can't
         # assign attributes to instance method values in Python 2
@@ -230,54 +237,6 @@ class GenericStateMachine(
         )
         state_machine_class._test_case_cache[state_machine_class] = StateMachineTestCase
         return StateMachineTestCase
-
-
-GenericStateMachine.find_breaking_runner = classmethod(find_breaking_runner)
-
-
-class StateMachineRunner(object):
-    """A StateMachineRunner is a description of how to run a state machine.
-
-    It contains values that it will use to shape the examples.
-    """
-
-    def __init__(self, data, n_steps):
-        self.data = data
-        self.data.is_find = False
-        self.n_steps = n_steps
-
-    def run(self, state_machine, print_steps=None):
-        if print_steps is None:
-            print_steps = current_verbosity() >= Verbosity.debug
-        self.data.hypothesis_runner = state_machine
-
-        should_continue = cu.many(
-            self.data, min_size=1, max_size=self.n_steps, average_size=self.n_steps
-        )
-
-        try:
-            if print_steps:
-                state_machine.print_start()
-            state_machine.check_invariants()
-
-            while should_continue.more():
-                value = self.data.draw(state_machine.steps())
-                if print_steps:
-                    state_machine.print_step(value)
-                state_machine.execute_step(value)
-                state_machine.check_invariants()
-        finally:
-            if print_steps:
-                state_machine.print_end()
-            state_machine.teardown()
-
-
-class StateMachineSearchStrategy(SearchStrategy):
-    def __init__(self, settings=None):
-        self.program_size = (settings or Settings.default).stateful_step_count
-
-    def do_draw(self, data):
-        return StateMachineRunner(data, self.program_size)
 
 
 @attr.s()
@@ -299,10 +258,10 @@ class Rule(object):
             else:
                 arguments[k] = v
         self.bundles = tuple(bundles)
-        self.arguments_strategy = fixed_dictionaries(arguments)
+        self.arguments_strategy = st.fixed_dictionaries(arguments)
 
 
-self_strategy = runner()
+self_strategy = st.runner()
 
 
 class BundleReferenceStrategy(SearchStrategy):
@@ -317,7 +276,7 @@ class BundleReferenceStrategy(SearchStrategy):
         # Shrink towards the right rather than the left. This makes it easier
         # to delete data generated earlier, as when the error is towards the
         # end there can be a lot of hard to remove padding.
-        return bundle[integer_range(data, 0, len(bundle) - 1, center=len(bundle))]
+        return bundle[cu.integer_range(data, 0, len(bundle) - 1, center=len(bundle))]
 
 
 class Bundle(SearchStrategy):
@@ -759,9 +718,9 @@ class RuleBasedStateMachine(GenericStateMachine):
     def steps(self):
         # Pick initialize rules first
         if self._initialize_rules_to_run:
-            return one_of(
+            return st.one_of(
                 [
-                    tuples(just(rule), fixed_dictionaries(rule.arguments))
+                    st.tuples(st.just(rule), st.fixed_dictionaries(rule.arguments))
                     for rule in self._initialize_rules_to_run
                 ]
             )

--- a/hypothesis-python/src/hypothesis/stateful.py
+++ b/hypothesis-python/src/hypothesis/stateful.py
@@ -121,6 +121,11 @@ def run_state_machine_as_test(state_machine_factory, settings=None):
                 machine.print_end()
             machine.teardown()
 
+    # Use a machine digest to identify stateful tests in the example database
+    run_state_machine.hypothesis.inner_test._hypothesis_internal_add_digest = function_digest(
+        state_machine_factory
+    )
+
     run_state_machine(state_machine_factory)
 
 

--- a/hypothesis-python/src/hypothesis/stateful.py
+++ b/hypothesis-python/src/hypothesis/stateful.py
@@ -125,6 +125,13 @@ def run_state_machine_as_test(state_machine_factory, settings=None):
     run_state_machine.hypothesis.inner_test._hypothesis_internal_add_digest = function_digest(
         state_machine_factory
     )
+    # Copy some attributes so @seed and @reproduce_failure "just work"
+    run_state_machine._hypothesis_internal_use_seed = getattr(
+        state_machine_factory, "_hypothesis_internal_use_seed", None
+    )
+    run_state_machine._hypothesis_internal_use_reproduce_failure = getattr(
+        state_machine_factory, "_hypothesis_internal_use_reproduce_failure", None
+    )
 
     run_state_machine(state_machine_factory)
 

--- a/hypothesis-python/src/hypothesis/strategies.py
+++ b/hypothesis-python/src/hypothesis/strategies.py
@@ -2184,13 +2184,13 @@ def runner(default=not_set):
 class DataObject(object):
     def __init__(self, data):
         self.count = 0
-        self.data = data
+        self.conjecture_data = data
 
     def __repr__(self):
         return "data(...)"
 
     def draw(self, strategy, label=None):
-        result = self.data.draw(strategy)
+        result = self.conjecture_data.draw(strategy)
         self.count += 1
         if label is not None:
             note("Draw %d (%s): %r" % (self.count, label, result))

--- a/hypothesis-python/tests/cover/test_stateful.py
+++ b/hypothesis-python/tests/cover/test_stateful.py
@@ -17,13 +17,12 @@
 
 from __future__ import absolute_import, division, print_function
 
-import inspect
 from collections import defaultdict, namedtuple
 
 import pytest
 from _pytest.outcomes import Failed, Skipped
 
-from hypothesis import assume, settings as Settings
+from hypothesis import PrintSettings, assume, seed, settings as Settings
 from hypothesis.control import current_build_context
 from hypothesis.database import ExampleDatabase
 from hypothesis.errors import Flaky, InvalidArgument, InvalidDefinition
@@ -50,6 +49,8 @@ from hypothesis.strategies import (
     tuples,
 )
 from tests.common.utils import capture_out, checks_deprecated_behaviour, raises
+
+NO_BLOB_SETTINGS = Settings(print_blob=PrintSettings.NEVER)
 
 
 class SetStateMachine(GenericStateMachine):
@@ -288,7 +289,7 @@ def test_multiple_rules_same_func():
 
 class GivenLikeStateMachine(GenericStateMachine):
     def steps(self):
-        return lists(booleans())
+        return lists(booleans(), min_size=1)
 
     def execute_step(self, step):
         assume(any(step))
@@ -323,10 +324,7 @@ class FlakyStateMachine(GenericStateMachine):
         return just(())
 
     def execute_step(self, step):
-        assert not any(
-            t[3] == u"find_breaking_runner"
-            for t in inspect.getouterframes(inspect.currentframe())
-        )
+        assert current_build_context().is_final
 
 
 def test_flaky_raises_flaky():
@@ -419,7 +417,7 @@ for test_case in (
     TestIntAdder,
     TestPrecondition,
 ):
-    test_case.settings = Settings(max_examples=10)
+    test_case.settings = Settings(test_case.settings, max_examples=10)
 
 
 def test_picks_up_settings_at_first_use_of_testcase():
@@ -451,25 +449,24 @@ def test_settings_are_independent():
 
 
 def test_minimizes_errors_in_teardown():
+    counter = [0]
+
     class Foo(GenericStateMachine):
         def __init__(self):
-            self.counter = 0
+            counter[0] = 0
 
         def steps(self):
             return tuples()
 
         def execute_step(self, value):
-            self.counter += 1
+            counter[0] += 1
 
         def teardown(self):
-            assert not self.counter
+            assert not counter[0]
 
-    runner = Foo.find_breaking_runner()
-
-    f = Foo()
     with raises(AssertionError):
-        runner.run(f, print_steps=True)
-    assert f.counter == 1
+        run_state_machine_as_test(Foo)
+    assert counter[0] == 1
 
 
 class RequiresInit(GenericStateMachine):
@@ -522,6 +519,26 @@ def test_can_explicitly_pass_settings():
         FailsEventually.TestCase.settings = Settings(
             FailsEventually.TestCase.settings, stateful_step_count=5
         )
+
+
+def test_settings_argument_is_validated():
+    with pytest.raises(InvalidArgument):
+        run_state_machine_as_test(FailsEventually, settings=object())
+
+
+def test_runner_that_checks_factory_produced_a_machine():
+    with pytest.raises(InvalidArgument):
+        run_state_machine_as_test(object)
+
+
+def test_settings_attribute_is_validated():
+    real_settings = FailsEventually.TestCase.settings
+    try:
+        FailsEventually.TestCase.settings = object()
+        with pytest.raises(InvalidArgument):
+            run_state_machine_as_test(FailsEventually)
+    finally:
+        FailsEventually.TestCase.settings = real_settings
 
 
 def test_saves_failing_example_in_database():
@@ -785,7 +802,7 @@ def test_prints_equal_values_with_correct_variable_name():
 
     result = o.getvalue()
     for m in ["create", "transfer", "fail"]:
-        assert result.count(m) == 1
+        assert result.count("state." + m) == 1
     assert "v1 = state.create()" in result
     assert "v2 = state.transfer(source=v1)" in result
     assert "state.fail(source=v2)" in result
@@ -816,7 +833,7 @@ def test_initialize_rule():
             run_state_machine_as_test(WithInitializeRules)
 
     assert set(WithInitializeRules.initialized[-3:]) == {"a", "b", "c"}
-    result = o.getvalue().splitlines()
+    result = o.getvalue().splitlines()[1:]
     assert result[0] == "state = WithInitializeRules()"
     # Initialize rules call order is shuffled
     assert {result[1], result[2], result[3]} == {
@@ -840,6 +857,7 @@ def test_initialize_rule_populate_bundle():
         def fail_fast(self, param):
             assert False
 
+    WithInitializeBundleRules.TestCase.settings = NO_BLOB_SETTINGS
     with capture_out() as o:
         with pytest.raises(AssertionError):
             run_state_machine_as_test(WithInitializeBundleRules)
@@ -847,7 +865,10 @@ def test_initialize_rule_populate_bundle():
     result = o.getvalue()
     assert (
         result
-        == """state = WithInitializeBundleRules()
+        == """\
+Falsifying example: run_state_machine(\
+factory=WithInitializeBundleRules, data=data(...))
+state = WithInitializeBundleRules()
 v1 = state.initialize_a(dep='dep')
 state.fail_fast(param=v1)
 state.teardown()
@@ -925,7 +946,7 @@ def test_initialize_rule_in_state_machine_with_inheritance():
             run_state_machine_as_test(ChildStateMachine)
 
     assert set(ChildStateMachine.initialized[-2:]) == {"a", "b"}
-    result = o.getvalue().splitlines()
+    result = o.getvalue().splitlines()[1:]
     assert result[0] == "state = ChildStateMachine()"
     # Initialize rules call order is shuffled
     assert {result[1], result[2]} == {"state.initialize_a()", "state.initialize_b()"}
@@ -944,8 +965,9 @@ def test_can_manually_call_initialize_rule():
 
         @rule()
         def fail_eventually(self):
-            assert self.initialize() == 2
+            assert self.initialize() <= 2
 
+    StateMachine.TestCase.settings = NO_BLOB_SETTINGS
     with capture_out() as o:
         with pytest.raises(AssertionError):
             run_state_machine_as_test(StateMachine)
@@ -953,7 +975,9 @@ def test_can_manually_call_initialize_rule():
     result = o.getvalue()
     assert (
         result
-        == """state = StateMachine()
+        == """\
+Falsifying example: run_state_machine(factory=StateMachine, data=data(...))
+state = StateMachine()
 state.initialize()
 state.fail_eventually()
 state.fail_eventually()
@@ -982,7 +1006,15 @@ def test_steps_printed_despite_pytest_fail(capsys):
     with pytest.raises(Failed):
         run_state_machine_as_test(RaisesProblem)
     out, _ = capsys.readouterr()
-    assert "state = RaisesProblem()\nstate.oops()\nstate.teardown()\n" == out
+    assert (
+        """\
+Falsifying example: run_state_machine(factory=RaisesProblem, data=data(...))
+state = RaisesProblem()
+state.oops()
+state.teardown()
+"""
+        in out
+    )
 
 
 def test_steps_not_printed_with_pytest_skip(capsys):
@@ -994,7 +1026,7 @@ def test_steps_not_printed_with_pytest_skip(capsys):
     with pytest.raises(Skipped):
         run_state_machine_as_test(RaisesProblem)
     out, _ = capsys.readouterr()
-    assert "" == out
+    assert "state" not in out
 
 
 @checks_deprecated_behaviour

--- a/hypothesis-python/tests/cover/test_stateful.py
+++ b/hypothesis-python/tests/cover/test_stateful.py
@@ -17,15 +17,23 @@
 
 from __future__ import absolute_import, division, print_function
 
+import base64
 from collections import defaultdict, namedtuple
 
 import pytest
 from _pytest.outcomes import Failed, Skipped
 
-from hypothesis import PrintSettings, assume, seed, settings as Settings
+from hypothesis import (
+    PrintSettings,
+    __version__,
+    assume,
+    reproduce_failure,
+    seed,
+    settings as Settings,
+)
 from hypothesis.control import current_build_context
 from hypothesis.database import ExampleDatabase
-from hypothesis.errors import Flaky, InvalidArgument, InvalidDefinition
+from hypothesis.errors import DidNotReproduce, Flaky, InvalidArgument, InvalidDefinition
 from hypothesis.internal.compat import print_unicode
 from hypothesis.stateful import (
     Bundle,
@@ -1051,3 +1059,38 @@ def test_rule_non_bundle_target_oneof():
     pattern = r".+ `one_of(a, b)` or `a | b` .+"
     with pytest.raises(InvalidArgument, match=pattern):
         rule(target=k | v)
+
+
+def test_uses_seed(capsys):
+    @seed(0)
+    class TrivialMachine(RuleBasedStateMachine):
+        @rule()
+        def oops(self):
+            assert False
+
+    with pytest.raises(AssertionError):
+        run_state_machine_as_test(TrivialMachine)
+    out, _ = capsys.readouterr()
+    assert "@seed" not in out
+
+
+def test_reproduce_failure_works():
+    @reproduce_failure(__version__, base64.b64encode(b"\0\0\0"))
+    class TrivialMachine(RuleBasedStateMachine):
+        @rule()
+        def oops(self):
+            assert False
+
+    with pytest.raises(AssertionError):
+        run_state_machine_as_test(TrivialMachine)
+
+
+def test_reproduce_failure_fails_if_no_error():
+    @reproduce_failure(__version__, base64.b64encode(b"\0\0\0"))
+    class TrivialMachine(RuleBasedStateMachine):
+        @rule()
+        def ok(self):
+            assert True
+
+    with pytest.raises(DidNotReproduce):
+        run_state_machine_as_test(TrivialMachine)

